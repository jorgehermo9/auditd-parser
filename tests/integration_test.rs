use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
