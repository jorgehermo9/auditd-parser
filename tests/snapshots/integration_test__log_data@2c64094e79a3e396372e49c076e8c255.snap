---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/auparse/test4.log:1"
expression: result
info: "type=SYSCALL msg=audit(1655465398.534:25618): arch=c000003e syscall=59 success=yes exit=0 a0=8c403a0 a1=8c3e8b0 a2=fffffb6cc5b0 a3=0 items=3 ppid=105182 pid=105183 auid=573 uid=583 gid=583 euid=583 suid=583 fsuid=583 egid=583 sgid=583 fsgid=583 tty=pts2 ses=2632 comm=\"ld\" exe=\"/bin/sh4\" key=(null)"
---
{
  "record_type": "SYSCALL",
  "timestamp": 1655465398534,
  "id": 25618,
  "fields": {
    "a0": "8c403a0",
    "a1": "8c3e8b0",
    "a2": "fffffb6cc5b0",
    "a3": "0",
    "arch": "x86_64",
    "auid": 573,
    "comm": "ld",
    "egid": 583,
    "euid": 583,
    "exe": "/bin/sh4",
    "exit": 0,
    "fsgid": 583,
    "fsuid": 583,
    "gid": 583,
    "items": "3",
    "key": null,
    "pid": "105183",
    "ppid": "105182",
    "ses": "2632",
    "sgid": 583,
    "success": true,
    "suid": 583,
    "syscall": "59",
    "tty": "pts2",
    "uid": 583
  }
}
